### PR TITLE
Add Cosine Similarity Loss for Distillation

### PR DESCRIPTION
**Pull request by Dexterai**
## Summary

This is a torch compiled, chunked fused linear Cosine Similarity Loss, aiming for knowledge distillation

**Consine Similarity Loss**

## Testing Done

Benchmarks
distill_cosine_loss_memory_full

distill_cosine_loss_speed_full

- Hardware Type: <BLANK>
- [x] run `make test` to ensure correctness
- [x] run `make checkstyle` to ensure code style
- [ ] run `make test-convergence` to ensure convergence

### CODE REVIEW

**Comment by Dexterai** [comment]
Could you please review this PR@Mecoli1219

File: src/liger_kernel/chunked_loss/cosine_similarity_loss.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            teacher_logits (torch.Tensor): Logits of teacher tokens. Shape: (batch_size * seq_len,).
+            temperature: Coefficient beta of generalized Cosine Similarity in the interval [0, 1]. Default: `1.0` (float): .
+        Returns:
+            torch.Tensor: cosine similarity loss
+        """
+        student_norm = F.normalize(student_logits, p=2, dim=-1)
+        teacher_norm = F.normalize(teacher_logits, p=2, dim=-1)
+
+        cosine_sim = F.cosine_similarity(student_norm, teacher_norm, dim=-1)
+        breakpoint()
```
Comment (shimizust) [inline src/liger_kernel/chunked_loss/cosine_similarity_loss.py]: Remove breakpoint

File: src/liger_kernel/chunked_loss/cosine_similarity_loss.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+class LigerFusedLinearCosineSimilarityFunction(LigerFusedLinearDistillationBase):
+    @staticmethod
+    def distillation_loss_fn(student_logits, teacher_logits, beta=1.0):
+        """
+        Compute Cosine loss (Cosine Similarity Loss).
+        Args:
+            student_logits (torch.Tensor): Logits of student tokens. Shape: (batch_size * seq_len,).
+            teacher_logits (torch.Tensor): Logits of teacher tokens. Shape: (batch_size * seq_len,).
+            temperature: Coefficient beta of generalized Cosine Similarity in the interval [0, 1]. Default: `1.0` (float): .
```
Comment (shimizust) [inline src/liger_kernel/chunked_loss/cosine_similarity_loss.py]: Should be beta, not temperature?

File: src/liger_kernel/chunked_loss/cosine_similarity_loss.py (hunk 1 above)
Comment (Dexterai) [reply src/liger_kernel/chunked_loss/cosine_similarity_loss.py]: Thanks for the advice, I will fix it right know

File: src/liger_kernel/chunked_loss/cosine_similarity_loss.py (hunk 2 above)
Comment (Dexterai) [reply src/liger_kernel/chunked_loss/cosine_similarity_loss.py]: That's right. Thanks for the advice.

**Review by shimizust — APPROVED** [review]
thank you for the contribution!
